### PR TITLE
feat: token api

### PR DESCRIPTION
**Pull request by giovanecosta**
[Rendered Design](https://github.com/HathorNetwork/hathor-explorer-service/blob/5068afcf7b3d3c39d1499a9f6efc6a645e5b8440/text/0006-token-api.md)

### CODE REVIEW

File: text/0006-token-api.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
+- Feature Name: token_api
+- Start Date: 2021-08-04
+- RFC PR: (leave this empty)
+- Hathor Issue: (leave this empty)
+- Author: Giovane Costa <gigio.coder@gmail.com>
+
+# Summary
+[summary]: #summary
+
+Token API will provide token information by requesting `hathor-wallet-service`.
```
Comment (msbrogli) [inline text/0006-token-api.md:10]: I guess we don't have to explain the low level in the summary. Maybe explain where it will be used.

File: text/0006-token-api.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+# Motivation
+[motivation]: #motivation
+
+We need to retrieve token data without requesting the full node. Wallet service holds those data, so, we'll use it for now.
```
Comment (msbrogli) [inline text/0006-token-api.md:15]: The motivation is to replace the full node's API so the explorer can scale as much as we want without hitting the full node direclty.

File: text/0006-token-api.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+             timestamp: int
+        }
+    ]
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+None.
```
Comment (msbrogli) [inline text/0006-token-api.md:82]: It depends on the wallet service.

File: text/0006-token-api.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+- We'll basically act as a proxy making requests to wallet service
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+**Endpoints:**
```
Comment (msbrogli) [inline text/0006-token-api.md:25]: I guess these endpoints belong to the guide level explanation.

Comment (msbrogli) [inline text/0006-token-api.md:25]: Maybe use `## Endpoints`.

File: text/0006-token-api.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+GET `/token/:id` return `200` | `404`
```
Comment (msbrogli) [inline text/0006-token-api.md:27]: Use `### GET ...`.

File: text/0006-token-api.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+A request is made for Wallet service passing the same id
```
Comment (msbrogli) [inline text/0006-token-api.md:29]: This explanation should be in the low level design. What is the API that will be accessed? Maybe add a link its documentation here.

File: tests/unit/usecases/test_list_tokens.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
+from unittest.mock import MagicMock
+
+from pytest import fixture
+
+from domain.tx.token import TokenList
+from tests.fixtures.token_factory import TokenListFactory
+from usecases.list_tokens import ListTokens
+
+
+class TestListTokens:
```
Comment (msbrogli) [inline tests/unit/usecases/test_list_tokens.py:10]: I don't feel we need tests as we are just proxying the response for the wallet service. What do you think?

File: tests/unit/usecases/test_list_token_transactions.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
+from unittest.mock import MagicMock
+
+from pytest import fixture
+
+from tests.fixtures.token_factory import TokenTransactionListFactory
+from usecases.list_token_transactions import ListTokenTransactions
+
+
+class TestListTokenTransactions:
```
Comment (msbrogli) [inline tests/unit/usecases/test_list_token_transactions.py:9]: As we are just doing a reverse proxy, do we still need tests?

File: tests/unit/gateways/clients/test_hathor_wallet_service_client.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    def test_get(self, mocked_get):
+        mocked_get.return_value.status_code = 200
+        mocked_get.return_value.json = MagicMock(return_value={'success': True})
+        mocked_get.return_value.raise_for_status = MagicMock()
+
+        client = HathorWalletServiceClient('mydomain.com')
+
+        result = client.get('/some/path', {'page': 2})
+
+        mocked_get.assert_called_once_with('https://mydomain.com/some/path', params={'page': 2})
```
Comment (msbrogli) [inline tests/unit/gateways/clients/test_hathor_wallet_service_client.py:21]: I like checking whether we are requesting the info from the wallet service.

File: tests/unit/gateways/clients/responses/hathor_wallet_service/test_hathor_wallet_service_token_response.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    HathorWalletServiceTokenTransactionsReponse,
+)
+from tests.fixtures.hathor_wallet_service_fixtures import (
+    HATHOR_WALLET_SERVICE_GET_TOKEN,
+    HATHOR_WALLET_SERVICE_GET_TOKEN_LIST,
+    HATHOR_WALLET_SERVICE_GET_TOKEN_TRANSACTION_LIST,
+)
+
+
+class TestHathorWalletServiceTokenResponse:
```
Comment (msbrogli) [inline tests/unit/gateways/clients/responses/hathor_wallet_service/test_hathor_wallet_service_token_response.py:16]: Same thing here. I don't know whether this test is useful since we are returning exactly the same thing as the wallet service.

File: tests/unit/domain/tx/test_token.py [hunk 11]
```diff
@@ -40,7 +40,6 @@ def test_token_to_dict(self):
         assert token_dict['name'] == token.name
         assert token_dict['symbol'] == token.symbol
         assert token_dict['total_supply'] == token.total_supply
-        assert token_dict['transactions_count'] == token.transactions_count
```
Comment (msbrogli) [inline tests/unit/domain/tx/test_token.py:43]: Why did you remove it?

File: tests/unit/domain/tx/test_token.py [hunk 12]
```diff
@@ -55,6 +54,5 @@ def test_token_from_dict(self):
         assert new_token.name == token.name
         assert new_token.symbol == token.symbol
         assert new_token.total_supply == token.total_supply
-        assert new_token.transactions_count == token.transactions_count
```
Comment (msbrogli) [inline tests/unit/domain/tx/test_token.py:58]: Why did you remove it?

File: serverless.yml [hunk 13]
```diff
@@ -121,6 +121,32 @@ functions:
...
+        - 'handlers/list_tokens_handler.py'
+    layers:
+      - { Ref: PythonRequirementsLambdaLayer }
+    events:
+      - http:
+          path: tokens/
+          method: get
+          cors: true
+  
+  list_token_transactions_handler:
```
Comment (msbrogli) [inline serverless.yml:138]: What is the timeout for this lambda? Should we use k8s instead?

File: serverless.yml [hunk 14]
```diff
@@ -121,6 +121,32 @@ functions:
             ttlInSeconds: 15
             cacheKeyParameters:
               - name: request.path.hash
+  
+  list_tokens_handler:
```
Comment (msbrogli) [inline serverless.yml:125]: What is the timeout for this lambda? Should we use k8s instead?

File: text/0006-token-api.md [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+[motivation]: #motivation
+
+We need to retrieve token data without requesting the full node. Wallet service holds those data, so, we'll use it for now. 
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+- We'll basically act as a proxy making requests to wallet service
+
+# Reference-level explanation
```
Comment (msbrogli) [inline text/0006-token-api.md:22]: I'm wondering whether we will just relay the response from the wallet service or not. I mean, should we have a well defined return type and map the Wallet Service's response into it? Or should we just be a reverse proxy and relay it?

I feel this is important to understand how this API might be affected by future changes in the Wallet Service's Token API.

File: handlers/list_tokens_handler.py [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+        'size': event.query.get('size', None)
+    }
+
+    response = list_token.list(params)
+
+    if response is None:
+        raise Exception('not_found')
+
+    return {
+        "statusCode": 200,
```
Comment (msbrogli) [inline handlers/list_tokens_handler.py:32]: What if `response.statusCode != 200`? Should we return 200 anyway?

File: handlers/list_token_transactions.py [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+        'size': event.query.get('size', None)
+    }
+
+    response = list_token_transactions.list(event.path['hash'], params)
+
+    if response is None:
+        raise Exception('not_found')
+
+    return {
+        "statusCode": 200,
```
Comment (msbrogli) [inline handlers/list_token_transactions.py:30]: What if `response.statusCode != 200`? Should we return 200 anyway?

File: gateways/token_gateway.py [hunk 18]
```diff
@@ -44,14 +56,85 @@ def get_token_metadata_from_s3(self, file: str) -> Union[TokenMetadata, None]:
...
+        :return: token domain class created from response
+        :rtype: Union[Token, None]
+        """
+        response = self.hathor_wallet_service_client.get(TOKEN_ENDPOINT.format(id=id), {})
 
         if response is None:
             return None
 
-        token_response = HathorCoreTokenResponse.from_dict(response)
+        token_response = HathorWalletServiceTokenResponse.from_dict(response)
```
Comment (msbrogli) [inline gateways/token_gateway.py:71]: I'm not sure whether we are just relaying the response or mapping to a well defined response type. In case we are relaying, I don't see why we would parse it.

File: gateways/clients/responses/hathor_wallet_service/hathor_wallet_service_token_response.py [hunk 19]
```diff
@@ -0,0 +1,161 @@
...
+            return TokenList(
+                total=self.total,
+                tokens=[TokenListItem(
+                    id=token.id,
+                    name=token.name,
+                    symbol=token.symbol,
+                    timestamp=token.timestamp
+                ) for token in self.tokens]
+            )
+        except Exception:
```
Comment (msbrogli) [inline gateways/clients/responses/hathor_wallet_service/hathor_wallet_service_token_response.py:160]: It feel dangerous to catch `Exception`. Why is that? It would capture ANY exception generated inside it (even if your code is broken).

File: gateways/clients/responses/hathor_wallet_service/hathor_wallet_service_token_response.py [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+            raise Exception('malformed_token_list')
```
Comment (msbrogli) [inline gateways/clients/responses/hathor_wallet_service/hathor_wallet_service_token_response.py:161]: Maybe raise a `ValueError`?

File: gateways/clients/responses/hathor_wallet_service/hathor_wallet_service_token_response.py [hunk 21]
```diff
@@ -0,0 +1,161 @@
...
+        """
+        try:
+            return TokenTransactionList(
+                total=self.total,
+                transactions=[TokenTransactionListItem(
+                    id=transaction.id,
+                    timestamp=transaction.timestamp
+                ) for transaction in self.transactions]
+            )
+        except Exception:
```
Comment (msbrogli) [inline gateways/clients/responses/hathor_wallet_service/hathor_wallet_service_token_response.py:102]: It feel dangerous to catch `Exception`. Why is that? It would capture ANY exception generated inside it (even if your code is broken).

File: gateways/clients/responses/hathor_wallet_service/hathor_wallet_service_token_response.py [hunk 22]
```diff
@@ -0,0 +1,161 @@
...
+            raise Exception('malformed_token_transaction_list')
```
Comment (msbrogli) [inline gateways/clients/responses/hathor_wallet_service/hathor_wallet_service_token_response.py:103]: Maybe raise `ValueError`?

File: gateways/clients/responses/hathor_wallet_service/hathor_wallet_service_token_response.py [hunk 23]
```diff
@@ -0,0 +1,161 @@
...
+    name: str
+    symbol: str
+    timestamp: int
+    total_supply: int
+    can_mint: bool
+    can_melt: bool
+
+    @classmethod
+    def from_dict(cls, dikt: dict) -> 'HathorWalletServiceTokenResponse':
+        return from_dict(data_class=cls, data=dikt)
```
Comment (msbrogli) [inline gateways/clients/responses/hathor_wallet_service/hathor_wallet_service_token_response.py:38]: Does it validate type?

File: text/0006-token-api.md (hunk 15 above)
Comment (msbrogli) [inline text/0006-token-api.md:22]: Still in the low level design, how will it be deployed? Lambdas, right? I'm concerned that it can get really expensive if the Wallet Service's Token API hangs for any reason and we keep waiting for them. What would be a reasonable timeout here? Should we deploy it using k8s?

**Review by msbrogli — CHANGES_REQUESTED** [review]
I reviewed the design and part of the code. I wasn't sure how to review the whole code since I still had some questions regarding the design.

File: domain/tx/token.py [hunk 24]
```diff
@@ -126,3 +122,84 @@ def from_dict(cls, dikt: dict) -> 'Token':
...
+    total: int
+    tokens: List[TokenListItem]
+
+    def to_dict(self) -> dict:
+        """ Convert a TokenList instance into dict
+
+        :return: Dict representations of TokenList
+        :rtype: dict
+        """
+        return asdict(self)
```
Comment (jansegre) [inline domain/tx/token.py:169]: I feel like all of these `to_dict` and `from_dict` aren't really helping in any way. I'd be fine using `asdict` and `from_dict` directly, or at most making thin wrapper functions on a util module in case we want to customize some behavior. What do you think?

File: gateways/clients/responses/hathor_wallet_service/hathor_wallet_service_token_response.py [hunk 25]
```diff
@@ -0,0 +1,161 @@
...
+
+from domain.tx.token import Token, TokenList, TokenListItem, TokenTransactionList, TokenTransactionListItem
+
+
+@dataclass
+class HathorWalletServiceTokenResponse:
+    """Response from hathor wallet service when requesting a token
+
+    :param id: id of token
+    :type id: str
```
Comment (jansegre) [inline gateways/clients/responses/hathor_wallet_service/hathor_wallet_service_token_response.py:14]: Are we generating a docs with these comments? If we are, is it possible to make it use the type annotations instead of having to duplicate the information with `:type`? If we aren't I feel like we have no feedback if we're making a mistake or not, and I wouldn't bother with `:type` at all, but the descriptions are nice.

File: text/0006-token-api.md [hunk 26]
```diff
@@ -0,0 +1,92 @@
+- Feature Name: token_api
+- Start Date: 2021-08-04
+- RFC PR: (leave this empty)
+- Hathor Issue: (leave this empty)
+- Author: Giovane Costa <gigio.coder@gmail.com>
```
Comment (jansegre) [inline text/0006-token-api.md:5]: Since this isn't in the RFC repo, I think we should remove this, because these fields are mostly specific to the RFC repo. And maybe also drop the `0006-` from the file name (I'm not sure if we have an RFC 0006 or not but it would certainly conflict with that).

**Review by jansegre — APPROVED** [review]
